hbt/production/btag.py

```diff
@@ -0,0 +1,323 @@
+            f"jes{'' if self.jec_source == 'Total' else self.jec_source}",
+            self.shift_inst.direction,
+            f"btag_weight_jec_{self.jec_source}_{self.shift_inst.direction}",
+        )
+    else:
+        # any other shift, just produce the nominal weight
+        events = add_weight("central", None, "btag_weight")
+
+    return events
+
+
+@btag_weight.init
+def btag_weight_init(self: Producer) -> None:
+    # depending on the requested shift_inst, there are three cases to handle:
+    #   1. when a JEC uncertainty is requested whose propagation to btag weights is known, the
+    #      producer should only produce that specific weight column
+    #   2. when the nominal shift is requested, the central weight and all variations related to the
+    #      method-intrinsic shifts are produced
+    #   3. when any other shift is requested, only create the central weight column
+    if getattr(self, "dataset_inst", None) is None or self.dataset_inst.is_data:
+        self.jec_source = None
+        self.shift_is_known_jec_source = None
+        self.btag_uncs = None
+        return
+
+    # to handle this efficiently in one spot, store jec information
+    self.jec_source = self.shift_inst.x.jec_source if self.shift_inst.has_tag("jec") else None
+    btag_sf_jec_source = "" if self.jec_source == "Total" else self.jec_source
+    self.shift_is_known_jec_source = (
+        self.jec_source and
+        btag_sf_jec_source in self.config_inst.x("btag_sf_jec_sources", [])
+    )
+
+    # save names of method-intrinsic uncertainties
+    self.btag_uncs = {
+        "hf": "hf",
+        "lf": "lf",
+        "hfstats1": "hfstats1_{year}",
+        "hfstats2": "hfstats2_{year}",
+        "lfstats1": "lfstats1_{year}",
+        "lfstats2": "lfstats2_{year}",
+        "cferr1": "cferr1",
+        "cferr2": "cferr2",
+    }
+
+    # add uncertainty sources of the method itself
+    if self.shift_inst.is_nominal:
+        # nominal column
+        self.produces.add("btag_weight")
+        # all varied columns
+        for col_name in self.btag_uncs.values():
+            name = col_name.format(year=self.config_inst.campaign.x.year)
+            for direction in ["up", "down"]:
+                self.produces.add(f"btag_weight_{name}_{direction}")
+    elif self.shift_is_known_jec_source:
+        # jec varied column
+        self.produces.add(f"btag_weight_jec_{self.jec_source}_{self.shift_inst.direction}")
+    else:
+        # only the nominal column
+        self.produces.add("btag_weight")
+
+
+@btag_weight.requires
+def btag_weight_requires(self: Producer, reqs: dict) -> None:
+    if self.dataset_inst.is_data or "external_files" in reqs:
+        return
+
+    from columnflow.tasks.external import BundleExternalFiles
+    reqs["external_files"] = BundleExternalFiles.req(self.task)
+
+
+@btag_weight.setup
+def btag_weight_setup(self: Producer, reqs: dict, inputs: dict) -> None:
+    if self.dataset_inst.is_data:
+        return
```

Comment:
Before `return`, `self.btag_sf_corrector` should be set to `None` to be consistent.